### ISSUE TITLE
学习率与batchsize关系

### ISSUE THREAD

**Issue by cszhou** [PaddlePaddle/Paddle#3453]
看了book里面的几个demo，有的学习率要除以batchsize，如`03.image_classification`
```Python
# Create optimizer
    momentum_optimizer = paddle.optimizer.Momentum(
        momentum=0.9,
        regularization=paddle.optimizer.L2Regularization(rate=0.0002 * 128),
        learning_rate=0.1 / 128.0,
        learning_rate_decay_a=0.1,
        learning_rate_decay_b=50000 * 100,
        learning_rate_schedule='discexp')
```
有的又没有除，如`06.understand_sentiment`
```Python
 # create optimizer
    adam_optimizer = paddle.optimizer.Adam(
        learning_rate=2e-3,
        regularization=paddle.optimizer.L2Regularization(rate=8e-4),
        model_average=paddle.optimizer.ModelAverage(average_window=0.5))
```
我的模型中，L2惩罚设置为0，只设置了全局的学习率, 没有对每一层再次设置学习率。在batchsize大小为30k的时候，发现学习率不管是设置为 0.001，还是0.001/30000的时候，都是4~5个batch（不是pass）就会稳定loss到0.3左右。
这里面有什么讲究？

**Comment by typhoonzero** [comment]
Not sure these might help:

Some general discussions: https://stats.stackexchange.com/questions/140811/how-large-should-the-batch-size-be-for-stochastic-gradient-descent

1. [The Effects of Hyperparameters on SGD Training of Neural Networks](https://arxiv.org/abs/1508.02788)
1. [On Large-Batch Training for Deep Learning: Generalization Gap and Sharp Minima](https://arxiv.org/abs/1609.04836)
1. [Systematic evaluation of CNN advances on the ImageNet](https://arxiv.org/abs/1606.02228)

**Comment by cszhou** [comment]
谢谢帮忙找超参设置的paper。
这里的一个困惑是，由于有的学习率除了batch size，有些没有，所以这里你们写代码的时候是将学习率与batch size解耦开，还是没有。
方便找下你们实现梯度下降的代码位置么。

**Comment by typhoonzero** [comment]
learning_rate和batch_size是解耦的，是单独的两个配置。

SGD实现可以参考源码：
入口：`paddle/trainer/Trainer.cpp`
实现：`paddle/gserver/gradientmachines/MultiGradientMachine.cpp`

**Comment by cszhou** [comment]
@typhoonzero 
Thanks

**Comment by lcy-seso** [comment]
在通常的教程/教科书可以看到如下公式：
- 损失函数通常有如下形式：
![image](https://user-images.githubusercontent.com/5842774/29260125-6eea8c24-80fa-11e7-8fda-6d853595a1f7.png)
- SGD 更新公式：
![image](https://user-images.githubusercontent.com/5842774/29260116-5c900996-80fa-11e7-97da-7885ac457aeb.png)

- Paddle中计算出的更新量不除以batch size，这个除以batch size的操作通过在学习率中得到补偿，但是在自适应学习率算法中，这和原更新公式并不是严格数学等价的。

**Comment by reyoung** [comment]
This issue has been answered
